src/flakeref/fr_type.rs

```diff
@@ -0,0 +1,387 @@
+use std::{
+    fmt::Display,
+    path::{Path, PathBuf},
+};
+
+use nom::{
+    branch::alt,
+    bytes::complete::{tag, take_till, take_until},
+    combinator::{opt, rest, verify},
+    IResult,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    error::{NixUriError, NixUriResult},
+    flakeref::forge::GitForge,
+    parser::parse_url_type,
+};
+
+use super::{GitForgePlatform, UrlType};
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum FlakeRefType {
+    // In URL form, the schema must be file+http://, file+https:// or file+file://. If the extension doesn’t correspond to a known archive format (as defined by the tarball fetcher), then the file+ prefix can be dropped.
+    File {
+        url: PathBuf,
+    },
+    /// Git repositories. The location of the repository is specified by the attribute
+    /// `url`. The `ref` arrribute defaults to resolving the `HEAD` reference.
+    /// The `rev` attribute must exist in the branch or tag specified by `ref`, defaults
+    /// to `ref`.
+    Git {
+        url: String,
+        r#type: UrlType,
+    },
+
+    GitForge(GitForge),
+    Indirect {
+        id: String,
+        ref_or_rev: Option<String>,
+    },
+    // Matches `git` type, but schema is one of the following:
+    // `hg+http`, `hg+https`, `hg+ssh` or `hg+file`.
+    Mercurial {
+        url: String,
+        r#type: UrlType,
+    },
+    /// Path must be a directory in the filesystem containing a `flake.nix`.
+    /// Path must be an absolute path.
+    Path {
+        path: String,
+    },
+    Tarball {
+        url: String,
+        r#type: UrlType,
+    },
+    #[default]
+    None,
+}
+impl FlakeRefType {
+    pub fn parse_file(input: &str) -> IResult<&str, &Path> {
+        let (rest, _) = tag("path:")(input)?;
+        let (rest, path_str) = verify(take_till(|c| c == '#' || c == '?'), |c: &str| {
```

Comment:
I think this verification is new? We want to make sure that we return a proper error, if possible on why the parsing failed.